services/user/Chainmail/service/src/lib.rs

```diff
@@ -121,8 +29,44 @@ mod service {
             .sent(get_sender(), receiver, subject, body);
     }
 
+    #[action]
+    fn archive(event_id: u64) {
+        Wrapper::emit()
+            .history()
+            .archive(get_sender().to_string() + &event_id.to_string());
+    }
+
+    #[action]
+    fn save(event_id: u64, sender: AccountNumber, subject: String, body: String) {
+        let saved_messages_table = SavedMessagesTable::new();
+
+        saved_messages_table
+            .put(&SavedMessage {
+                event_id,
+                sender,
+                subject,
+                body,
+            })
+            .unwrap();
+        ()
+    }
+
+    #[action]
+    fn unsave(event_id: u64, sender: AccountNumber, subject: String, body: String) {
+        let saved_messages_table = SavedMessagesTable::new();
+
+        saved_messages_table.remove(&SavedMessage {
+            event_id,
+            sender,
+            subject,
+            body,
+        })
+    }
+
     #[event(history)]
     pub fn sent(sender: AccountNumber, receiver: AccountNumber, subject: String, body: String) {}
+    #[event(history)]
+    pub fn archive(event_id: String) {}
```

Comment:
If you add an email id field, you can make this `archive(account: AccountNumber, email_id: u64)`